vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ Copyright 2018 New Vector Ltd
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package im.vector.analytics.e2e
+
+import im.vector.analytics.Analytics
+import im.vector.analytics.TrackingEvent
+import org.matrix.androidsdk.crypto.MXCryptoError
+import org.matrix.androidsdk.data.RoomState
+import org.matrix.androidsdk.listeners.MXEventListener
+import org.matrix.androidsdk.rest.model.Event
+import org.matrix.androidsdk.rest.model.RoomMember
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.collections.ArrayList
+import kotlin.concurrent.timer
+
+/**
+ * This class is responsible for managing the reported decryption errors
+ */
+class DecryptionFailureTracker(
+        private val analytics: Analytics) : MXEventListener() {
+
+    private val reportedFailures = ConcurrentHashMap<String, DecryptionFailure>()
+    private val trackedEvents = HashSet<String>()
+    private val checkFailuresTimer = timer(period = CHECK_PERIOD) {
+        checkFailures()
+    }
+
+    /**
+     * Reports the decryption error to the tracker.
+     * The error will be filtered.
+     * @param event: The error event
+     * @param roomState: The state of the room when the decryption error happened
+     * @param userId: The user id of the current session user
+     */
+    fun reportUnableToDecryptError(event: Event, roomState: RoomState, userId: String) {
+        if (reportedFailures[event.eventId] != null || trackedEvents.contains(event.eventId)) {
+            return
+        }
+        // Filter out "expected" UTDs
+        // We cannot decrypt messages sent before the user joined the room
+        val myUser = roomState.getMember(userId)
+        if (myUser.membership != RoomMember.MEMBERSHIP_JOIN) {
+            return
```

Comment:
Maybe we should report this type of error with a specific DecryptionFailureReason type? @manuroe ?

Reply:
N/A (discussed IRL)